server/src/main/java/org/elasticsearch/plugins/Plugin.java

```diff
@@ -97,7 +99,8 @@ public Collection<Object> createComponents(
         IndexNameExpressionResolver indexNameExpressionResolver,
         Supplier<RepositoriesService> repositoriesServiceSupplier,
         Tracer tracer,
-        AllocationService allocationService
+        AllocationService allocationService,
+        IndicesService indicesService
```

Comment:
Actual change is here. 

I wonder if it is worth extracting all above parameters into an object (PluginDependencies? ElasticsearchDependencies? Parameters)?
This way adding a new service would only touch the object and its creation rather then all places where this API is used.

Reply:
I think we should involve the @elastic/es-core-infra team if we're planning to do that change?

Reply:
This really feels like a place where some kind of dependency injection would be a better solution. Almost every plugin ignores almost every one of these dependencies, and yet rather often we encounter cases where the list of dependencies needs just one more thing adding.

That said, if we need this specific dependency then IMO we can add it without core/infra approval, but it would be great if the core/infra folks could come up with a better solution here (/cc @pgomulka with whom I was discussing this the other day).

Reply:
> That said, if we need this specific dependency then IMO we can add it without core/infra approval, but it would be great if the core/infra folks could come up with a better solution here (/cc @pgomulka with whom I was discussing this the other day).

Sorry if my comment wasn't clear enough, I meant that we should involve core/infra if we're planning to change how we plug dependencies for plugins.

Reply:
We definitely should; we're going to have to add SearchService and IndicesService to the list as well, so all the service objects should be in a separate POJO passed to the plugins here

Reply:
++ to some kind of Parameter Object refactoring. 
we could perhaps refactor this more into ActionPlugin::getActions style which allows to inject dependencies into transport actions.
however
@rjernst mentioned that there was actually work in the past to steer away from this. Can you give us more details on why we are reluctant to do DI in ES codebase?

Reply:
I agree to David here in that we should proceed with this without introducing a parameter object, leaving that to the core/infra team as a separate effort.

Reply:
> it would be great if the core/infra folks could come up with a better solution here

Improving how services get dependencies is something we would very much like to do. Unfortunately it is not a small effort, and other work currently takes priority. For now, adding to the createComponents signatures is the only way.

> if we need this specific dependency then IMO we can add it without core/infra approval

One reason we moved away from DI in plugin initialization is that dependencies were added without thought, leading to a mess of various levels of random objects/services/etc available, and needing to be initialized in a difficult to understand and control order. The services currently available in createComponents are carefully constructed by hand, and the initialization order is controlled within Node. We need to be careful about which things are added, especially if we hope to re-add some DI to make this a bit easier. So please do include core/infra with any changes to plugin dependencies, we need to be aware of these both for keeping the init order in Node sane, as well as for ensuring future ideas for plugin init are not made more difficult by quick addition of additional dependencies to these params.

Reply:
I want to add there may be one other model possible that does not require changing createComponents here. It depends on exactly what is to be done with IndicesService. We have other examples where we have dedicated plugin interfaces for additional functional (and since we are now modularized, we can limit access to those plugin interfaces to only ES modules).

For example, the IngestService is currently not needed in createComponents because functionality that extends ingest is handled through IngestPlugin and these are loaded and passed to IngestService by Node. So another option might be to add a new plugin interface with a hook for the IndicesService to pull whatever additional functionality is desired. It won't work in all circumstances, but it is something to consider.